internal/storage/chunk_uploader.go

```diff
@@ -0,0 +1,193 @@
+	totalWriteSucceededSoFar atomic.Int64
+	userProgressFunc         func(int64)
+
+	// Internal state for lifecycle management.
+	state UploaderState
+}
+
+// NewChunkUploader creates a new instance of chunkUploader,
+// for the given inputs.
+func NewChunkUploader(ctx context.Context, obj *storage.ObjectHandle, req *gcs.CreateObjectRequest, writeChunkSize int, progressFunc func(int64)) (gcs.ChunkUploader, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("ctx is nil")
+	}
+	if obj == nil || req == nil {
+		return nil, fmt.Errorf("nil ObjectHandle or CreateObjectRequest")
+	}
+
+	if obj.ObjectName() != req.Name {
+		return nil, fmt.Errorf("names of passed ObjectHandle and CreateObjectRequest don't match: ObjectHandle.Name=%s createObjectRequest.Name=%s", obj.ObjectName(), req.Name)
+	}
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		return nil, fmt.Errorf("request received for pre-existing object %s, supported only for new objects", req.Name)
+	}
+
+	if writeChunkSize <= 0 {
+		return nil, fmt.Errorf("chunkSize <= 0")
+	}
+
+	// Raw initialization.
+	uploader := chunkUploader{}
+
+	// Store references to necessary parameters.
+	uploader.objectName = obj.ObjectName()
+	uploader.userProgressFunc = progressFunc
+
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// NewWriter never returns nil, so no nil-check is needed on it.
+	wc := obj.NewWriter(ctx)
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = writeChunkSize
+	wc.ProgressFunc = func(n int64) {
+		uploader.totalWriteSucceededSoFar.Store(n)
+		logger.Debugf("%d bytes copied so far for object/file %s. chunk-size = %d", n, req.Name, wc.ChunkSize)
+
+		if uploader.userProgressFunc != nil {
+			uploader.userProgressFunc(n)
+		}
+	}
+
+	uploader.writer = wc
+	uploader.state = Initialized
+
+	return &uploader, nil
+}
+
+// BytesUploadedSoFar returns the total number of bytes successfully uploaded
+// so far using this uploader.
+// This is thread-safe against the
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+func (uploader *chunkUploader) BytesUploadedSoFar() int64 {
+	return uploader.totalWriteSucceededSoFar.Load()
+}
+
+func (uploader *chunkUploader) readyToUpload() bool {
+	switch uploader.state {
+	case Initialized, Uploading:
+		return true
+	default:
+		return false
+	}
+}
+
+func (uploader *chunkUploader) readyToClose() bool {
+	switch uploader.state {
+	case Initialized, Uploading, UploadError:
+		return true
+	default:
+		return false
+	}
+}
+
+// UploadChunkAsync uploads the passed content to GCS.
+// This waits (using mutex) until the completion of
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+// The call times out if it is taking longer than StorageCopyTimeoutDuration.
+func (uploader *chunkUploader) UploadChunkAsync(contents io.Reader) error {
```

Comment:
The function is asynchronous as it queues up the GCS uploads and returns right after queueing up the last chunk provided in the content, rather than blocking on that chunk-upload finishing. When the uploaded content is a single chunk (which is the main use case), then it just queues up the gcs upload and returns. There will not be much value to calling this in a goroutine if we stick to single-chunk uploads.

Reply:
When the buffer size is greater than 50 MiB, chunk uploader will get `contents` which has more than 1 chunk size of data. So in that case, this will be a blocking call for caller. Should we do something to handle this ?

Reply:
@sethiay As discussed offline, if the uploader has to be used for uploading multiple chunks, then this function should be called in a goroutine on the caller side. 
I'll add the context for timing out the call in a goroutine as you suggested though.

Reply:
Done.